### PR TITLE
ENH: allow better categorical dtype strings, e.g. `category[string]`

### LINKED ISSUE #48515: ENH: Improved `CategoricalDtype` subtype handling.
### Feature Type

- [X] Adding new functionality to pandas

- [X] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

Internally categories already distinguish different subtypes. consider for example:

```python
import pandas as pd
s = pd.Series(["foo", "bar"], dtype=object)
print(s.astype("category"))
print(s.astype("string").astype("category"))
```

In the first case, `s.dtype.categories` is `Index(['bar', 'foo'], dtype='object')`, in the latter case it is `Index(['bar', 'foo'], dtype='string')`.

However currently handling of these subtypes is a bit awkward, hence the proposed features are quality-of-life improvements when working with such kinds of data, mainly:

1. Allow direct casting to categories of specific subtype via `.astype("category[<type>]")`
2. Ensure round tripping subtypes when serializing in formats that support categorical types.

```python
import pandas as pd
df = pd.DataFrame({"col":[ "foo", "bar"]}, dtype=object)
df = df.astype("string").astype("category")
df.to_parquet("test.parquet")
print(df["col"].dtype.categories)
df= pd.read_parquet("test.parquet")
print(df["col"].dtype.categories)
```

### Feature Description

- [ ] Make `CategoricalDtype` a `typing.Generic` parametrized by a scalar type. (⇝ relevant for `pandas-stubs`)
- [ ] The fallback should be `category[object]` (cf. https://github.com/python/mypy/issues/4236#issuecomment-521628880)
- [ ] Allow `.astype("category[<type>]")`, e.g. `series.astype("category[string]")` should behave equivalently to `series.astype("string").astype("category")`
- [ ] Allow `category[<type>]` in `read_csv(file, dtype=...)`
- [ ] Ensure category subtypes are maintained trough serialization and loading
   - In particular, when reading parquet/feather format. (⇝ interoperability with `pyarrow`'s  dictionary type)

### Alternative Solutions

Existing functionality is to manually cast as `.astype(<type>).astype("category")` whenever necessary.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by topper-123**
- [X] closes #48515
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This PR is a quality-of-life improvement when working with categorical dtypes:

```python
>>> df = pd.read_csv(file, dtypes={"a": "category[string]", "b": "category[datetime64[ns, UTC]]"} 
```

It can also be visible in reprs and can be used for comparisons:

```python
>>> ser = pd.Series([1,2,3, 4], dtype="category[Int8]")
>>> ser
0    1
1    2
2    3
3    4
dtype: category[Int8]
Categories (4, Int8): [1, 2, 3, 4]
>>> ser.dtype == "category"
True
>>> ser.dtype == "category[Int8]"
True
>>> ser.dtype == "category[string]"
False
```

Still needs to do
- [ ]  update the tests.

EDIT: updated to newest capabilities.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
i was expecting this to be about getting a more informative string when doing `df.dtypes`.

I like the added specificity, but dont like adding more statefulness to the Dtype object.  once the categories are known the extra state should be redundant right?

**Comment by topper-123** [comment]
> i was expecting this to be about getting a more informative string when doing `df.dtypes`.

I was actually thinking about doing that as a follow-up, so we'd get e.g.:

```python
>>> cat = pd.Categorical(["a", "f"])
>>> pd.DataFrame(cat).dtypes
0    category[object]
dtype: object
```

> I like the added specificity, but dont like adding more statefulness to the Dtype object. once the categories are known the extra state should be redundant right?

I could not initiate `_categories_dtype` as None + delete it after the categories have been declared?

The issue is that `CategoricalDtype` can be instantiated without categories, so to make "category[string]" work as a string dtype, we have to store the dtype string somewhere, until the categories are declared.

**Comment by topper-123** [comment]
I've updated this, so in addition to being able to use these strings as dtypes in constructors, they will also be visible in reprs and can be used for comparisons:

```python
>>> ser = pd.Series([1,2,3, 4], dtype="category[Int8]")
>>> ser
0    1
1    2
2    3
3    4
dtype: category[Int8]
Categories (4, Int8): [1, 2, 3, 4]
>>> ser.dtype == "category"
True
>>> ser.dtype == "category[Int8]"
True
>>> ser.dtype == "category[string]"
False
```
 I still need to do the tests and before I do them, can you verify you approve of the idea, @jbrockmendel?

File: pandas/core/dtypes/dtypes.py [hunk 1]
```diff
@@ -152,13 +152,17 @@ class CategoricalDtype(PandasExtensionDtype, ExtensionDtype):
...
-        The categories are stored in an Index,
-        and if an index is provided the dtype of that index will be used.
+        The categories are stored in an Index.
     ordered : bool or None, default False
         Whether or not this categorical is treated as a ordered categorical.
         None can be used to maintain the ordered value of existing categoricals when
         used in operations that combine categoricals, e.g. astype, and will resolve to
         False if there is no existing ordered to maintain.
+    categories_dtype : dtype, optional
+        If given, will be the dtype of the categories.
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:162]: dtype is a better name

**Review by jreback — CHANGES_REQUESTED** [review]
i am surprised this doesn't break a lot of things

i like the repr idea but i am not sure allowing a dtype that is not inferred is actually useful

can you show when this is the case?

File: pandas/core/dtypes/dtypes.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/dtypes/dtypes.py:162]: I’d be ok to change, categories_dtype is very verbose.

**Comment by jbrockmendel** [comment]
i very much like the `__repr__` and `__eq__` here.  im not sold on it being in the constructor

**Comment by topper-123** [comment]
I'm not super fond of a categorical dtype repr not being usable as a dtype, i.e. if we allow `"category[string]" as a dtype repr, IMO not allow it as a dtype could be unexpected.

The `categories` attribute can currently be either None or a sequence. Could an idea be that `categories` could also be a dtype to represent an intermediate state, i.e. no sequence has yet been added, but we've restricted it to be of the given dtype?

For example `Series([1, 2, 3], dtype="category[string]")` would then be equivalent to `Series([1, 2, 3], dtype=CategoryDtype(StringDtype()))`? This would then resolve into `Series([1, 2, 3], dtype=CategoryDtype(Index([1, 2, 3], dtype="string")))` when `CategoricalDtype` has been given the sequence?

**Comment by mroeschke** [comment]
Appears the discussion here has stalled and might need a bit more discussion on the design which may good to have in the issue so closing for now. Happy to reopen at a later point

**Pull request closed without merging**